azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java

```diff
@@ -0,0 +1,113 @@
+package azkaban.hive;
+
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.hive.metastore.api.hive_metastoreConstants.META_TABLE_STORAGE;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.pool2.BasePooledObjectFactory;
+import org.apache.commons.pool2.PooledObject;
+import org.apache.commons.pool2.impl.DefaultPooledObject;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaHookLoader;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.metadata.HiveStorageHandler;
+import org.apache.hadoop.hive.ql.metadata.HiveUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link BasePooledObjectFactory} for {@link IMetaStoreClient}.
+ */
+public class HiveMetaStoreClientFactory extends BasePooledObjectFactory<IMetaStoreClient> {
+
+  public static final String HIVE_METASTORE_TOKEN_SIGNATURE = "hive.metastore.token.signature";
+  private static final Logger LOG = LoggerFactory.getLogger(HiveMetaStoreClientFactory.class);
+  private final HiveConf hiveConf;
+
+  public HiveMetaStoreClientFactory() {
+    this(new HiveConf());
+  }
+
+  public HiveMetaStoreClientFactory(final String hcatUri) {
+    this(getHiveConf(hcatUri));
+  }
+
+  public HiveMetaStoreClientFactory(final HiveConf hiveConf) {
+    this.hiveConf = hiveConf;
+  }
+
+  private static HiveConf getHiveConf(final String hcatURI) {
+    requireNonNull(hcatURI);
+    checkArgument(StringUtils.isNotBlank(hcatURI));
+
+    final HiveConf hiveConf = new HiveConf();
+    hiveConf.setVar(HiveConf.ConfVars.METASTOREURIS, hcatURI);
+    hiveConf.set(HIVE_METASTORE_TOKEN_SIGNATURE, hcatURI);
+
+    return hiveConf;
+  }
+
+  private IMetaStoreClient createMetaStoreClient() throws MetaException {
+    final HiveMetaHookLoader hookLoader = tbl -> {
+      if (tbl == null) {
+        return null;
+      }
+
+      try {
+        final HiveStorageHandler storageHandler =
+            HiveUtils.getStorageHandler(HiveMetaStoreClientFactory.this.hiveConf,
+                tbl.getParameters().get(META_TABLE_STORAGE));
+        return storageHandler == null ? null : storageHandler.getMetaHook();
+      } catch (final HiveException e) {
+        LOG.error(e.toString());
+        throw new MetaException("Failed to get storage handler: " + e);
+      }
+    };
+
+    return RetryingMetaStoreClient
+        .getProxy(this.hiveConf, hookLoader, HiveMetaStoreClient.class.getName());
+  }
+
+  @Override
+  public IMetaStoreClient create() {
+    try {
+      return createMetaStoreClient();
+    } catch (final MetaException e) {
```

Comment:
The `HiveException` is inside a Lambda. I don't think you can propagate that directly. Correct me if I'm wrong here.